File: src/TinRoll.Test/Manager/UserManagerTests.cs
using Moq;
using System;
using System.Collections.Generic;
using System.Text;
using TinRoll.Data.Entities;
using TinRoll.Data.Repository.Interface;
using TinRoll.Logic.Managers;
using TinRoll.Shared;
using Xunit;

namespace TinRoll.Test.Manager
{
    public class UserManagerTests
    {
        [Fact]
        public void Test_CreateUser()
        {
            //dummy user to return
            var mockUser = new User
            {
                Id = 1,
                Email = "test@email.com",
                UserName = "UserName"
            };
            

            //create a mock version of the IUserRepository
            var mockUserRepo = new Mock<IUserRepository>();

            //define what we want to happen when the CreateUserAsync method is called
            mockUserRepo.Setup(u => u.CreateUserAsync(It.IsAny<User>()))
                .ReturnsAsync(mockUser);

            //create UserManager with that Mock object
            var userManager = new UserManager(mockUserRepo.Object);

            //perform unit test
            var userToTest = new UserDto
            {
                Email = "test@email.com",
                UserName = "UserName"
            };
            var createdUser = userManager.CreateUserAsync(userToTest);

            //Assert
            Assert.Equal(1, createdUser.Id);
            mockUserRepo.Verify(u => u.CreateUserAsync(It.IsAny<User>()), Times.Once);
        }
    }
}

File: src/TinRoll.Test/Logic/Manager/UserManagerTests.fs
module Logic.Manager.UserManagerTests

open Xunit
open NFluent
open TinRoll.Logic.Manager
open TinRoll.Data
open System
open Moq
open TinRoll.Data.Repository
open TinRoll.Shared
open System.Collections

let GetTestDbUser () =
    new User (1, "terst@gail.com", "TestUsername", new Generic.List<Question>(), DateTime.MinValue, DateTime.MinValue)
let GetTestDtoUser () =
    new UserDto (Id = 1, Email = "tesrst@gmail.com", Username = "Testusername", Questions = new Generic.List<QuestionDto>(), CreatedDate = DateTime.MinValue, UpdatedDate = DateTime.MinValue)

[<Fact>]
let ``Test Get Users`` () =
    let dbUsers = async {
        return [GetTestDbUser()]
    }
    let mock = Mock<IUserRepository>();
    mock.Setup(fun ur -> ur.GetUsersAsync()).Returns(dbUsers) |> ignore

    let userManager = new UserManager(mock.Object) :> IUserManager
    let users = 
        userManager.GetUsersAsync()
        |> Async.AwaitTask
        |> Async.RunSynchronously


    Check.That(users.Length).IsEqualTo(1) |> ignore
    Check.That(users.Head.Id).IsEqualTo(1) |> ignore

[<Fact>]
let ``Test Get User`` () =
    let dbUser = async {
        return Some (GetTestDbUser())
    }

    let mock = Mock<IUserRepository>();
    mock.Setup(fun userRepo -> userRepo.GetUserAsync(It.IsAny<int>())).Returns(dbUser) |> ignore

    let userManager = new UserManager(mock.Object) :> IUserManager
    let user = 
        userManager.GetUserAsync 1
        |> Async.AwaitTask
        |> Async.RunSynchronously

    Check.That(user).IsNotNull() |> ignore
    Check.That(user.Id).IsEqualTo(1) |> ignore

[<Fact>]
let ``Test Create User`` () =
    let dtoUser = GetTestDtoUser()

    let userId = async {
        return 1
    }

    let mock = Mock<IUserRepository>();
    mock.Setup(fun userRepo -> userRepo.CreateUserAsync(It.IsAny<User>())).Returns(userId) |> ignore
    
    let userManager = new UserManager(mock.Object) :> IUserManager
    let userId = 
        userManager.CreateUserAsync dtoUser
        |> Async.AwaitTask
        |> Async.RunSynchronously
    
    Check.That(userId).IsEqualTo(1) |> ignore